### PR TITLE
First draft for sharing new ownership of code sections in Meteor.

### PR DESCRIPTION
**Pull request by queso**
For @stubailo, per sharing via https://forums.meteor.com/t/how-open-source-meteor-really-is-if-pull-requests-are-not-getting-in/16483/63.

A small start for a Friday afternoon :)

Sign offs to merge this PR:
- [x] @benjamn 
- [x] @stubailo 
- [x] @avital 
- [x] @yyx990803 
- [x] @martijnwalraven 
- [x] @Urigo

### CODE REVIEW

File: Contributing.md [hunk 1]
```diff
@@ -242,6 +242,14 @@ these guidelines:
  * Be sure your author field in git is properly filled out with your full name
  and email address so we can credit you.
 
+#### Need help with your pull request?
+
+Meteor now has groups defined to cover areas of the codebase. If you need help on certain topics you can address:
+
+* Meteor Data Team - This includes DDP, tracker, mongo, accounts, etc. You can talk to @stubailo.
+* Blaze - This includes Spacebars, Blaze, etc. You can talk to @urigo or @yyx990803.
+* Build tools - This includes modules, build tool changes, etc. You can talk to @avital or @benjamn.
```
Comment (mitar) [inline Contributing.md]: +1

It would be great if now a community member could just merge this in. :-)

Comment (mitar) [inline Contributing.md]: Maybe make links to their GitHub pages?

And also maybe place "talk" with how exactly you should talk. Like, no private messages probably.

Comment (queso) [inline Contributing.md]: I tweaked the wording now, thx.

**Comment by mitar** [comment]
Thanks to you!

+1

**Comment by RobertLowe** [comment]
Yea, that's super helpful, but I also think MDG needs to empower those people with final say on pulls so we don't end up in a design-by-committee situation.

**Comment by stubailo** [comment]
@benjamn @avital @martijnwalraven @yyx990803 @Urigo what do you think of this?

**Comment by benstr** [comment]
:+1:

**Comment by wearhere** [comment]
:+1:

**Comment by stubailo** [comment]
I'd like to get a sign-off on every single person mentioned here before this is merged, btw. Otherwise it seems unfair to enshrine their name in this document without their approval.

**Comment by mitar** [comment]
Or, they can just remove them, if they do not like it? :-) (To show an example of "let's perfect not be the enemy of the good" and "ask for forgiveness, not permission" which might speed up pull requests in general. We have version control for a reason. You can always revert things later on.)

**Comment by benstr** [comment]
oh, great point @mitar !

**Comment by benjamn** [comment]
signing off :lock_with_ink_pen:

**Comment by stubailo** [comment]
@mitar let's give it a couple days before we declare it "good". Say, until Wednesday. I'll track down every person by then and see what they think.

**Comment by mitar** [comment]
Let' see what happens if you do not track them personally. ;-) To see how long will it take to get progress on this pull request with normal processes. :-)

**Comment by martijnwalraven** [comment]
I definitely agree we should make it easier for pull requests to reach the right people, but I don't want this to make promises we can't keep. There is no such thing as offering unqualified help. So ideally it would be part of a larger effort to empower contributors and improve the project evolution process.

At a minimum, we should update [Contributing.md](https://github.com/meteor/meteor/blob/devel/Contributing.md) and link to it here.

One thing I think is important to acknowledge is that there are different kinds of pull requests, and these will always be treated differently. Merging bug fixes still takes effort and may require some back and forth, but should be a relatively straightforward process.

Proposing new features however, will almost never just be a matter of merging a pull request. It requires more careful discussion, preferably beforehand so fresh ideas can still influence the design direction. And there is no guarantee new features will be accepted. I'm not sure pull requests are the right medium for this.

This may be too heavyweight for us, but I really admire the [Swift Evolution Process](https://github.com/apple/swift-evolution/blob/master/process.md). It asks people to take the time to write [a proposal](https://github.com/apple/swift-evolution/blob/master/0000-template.md) to motivate a feature and describe the proposed design, but in turn promises that these proposals will be reviewed by someone from the core team. A similar mechanism, where we offer contributors more guidance but also ask that they take the time to communicate their ideas well and adhere to a process might be helpful here.

**Comment by Urigo** [comment]
I'm very excited about that! signing off!

Few thoughts though:

1. The easiest way to make the process of merging PRs faster is to improve our repo's tooling - Labels and Milestones (we are doing great with 1.3 Github milestones), commit message convention tools and most importantly - make our **Travis integration** reliable and add **tests** with **test coverage tools** - I want to see help from the community on those and I will make those first priority PRs. feel free to tag me on those. We can learn and copy from the Angular repos on that front.
2. @martijnwalraven, the proposal process looks really nice but I think we shouldn't require it.
Even on large new features, I find it much more productive to have the discussion around an actual code and not a conceptual doc, it can save us all a lot of time.

**Comment by stubailo** [comment]
I guess I feel really bad when someone has already written a lot of code but it's a feature we shouldn't accept. I'd rather we had discussions about the design in general, not a specific implementation, which is what happens when a proposal includes code.

**Comment by mitar** [comment]
> I guess I feel really bad when someone has already written a lot of code but it's a feature we shouldn't accept. I'd rather we had discussions about the design in general, not a specific implementation, which is what happens when a proposal includes code.

Probably this is one of the main reasons why I would like to see more community governance. node.js had exactly the same issue. Many features wanted by community was not accepted into node.js. This is really problematic. In the radio show you said that you care about companies who depend on the Meteor and what features they need. But there are also one other group of people you should care: developers of libraries for Meteor who might want to live from providing those libraries. With your approach of accepting features it is really hard for them. But on the other hand, those are exactly the developers who you would want as well: those who depend on Meteor for their income.

Anyway, I think features should also be merged with less complications. In the sense: do they have tests, do they have documentation, are they backwards compatible, do some people need this feature? Yes, merged. If later on people who depend on this feature disappear and stop maintaining it, then it is not a problem to remove it, they do not care about it anyway. If there are people who do depend on it, then great, you are solving for some people some need so keep it in.

**Comment by martijnwalraven** [comment]
> If later on people who depend on this feature disappear and stop maintaining it, then it is not a problem to remove it, they do not care about it anyway.

This is where is disagree. In practice, features are almost never removed because it is impossible to know when they are no longer used. They may also have affected the design in ways that make it hard to remove them (and that may have complicated things unnecessarily).

**Comment by mitar** [comment]
> In practice, features are almost never removed because it is impossible to know when they are no longer used.

Hm, using static analysis of all packages on Atmosphere and seeing what parts of Meteor they use, and how many installations of those packages are? You could get candidates for feature removal. And then you deprecate them in one version, see how many people complain, and then remove in the next. (For paying private code you have Galaxy and can probably do something similar there as well.)

We could do some big data crunching to improve this. Because I do think that otherwise every project is just growing and growing and gets unmaintainable. You should have a process of deletion as well. Like, if not enough people are using a feature, delete it. The process could also be manual. Every so and so MDG while working on the code finds parts of it which would like to remove. They put this on some website pending feedback from people. People can then explain why it should not be removed. If there is nobody who claims use stories at later time, it goes into deprecation and later on into a removal.

**Comment by martijnwalraven** [comment]
I don't think static analysis or other automated methods are a viable option for this. That is not just an issue of practicality, but I believe there are good conceptual reasons why this is at its core a social problem.

Once merged, most features are not a recognizable independent part of the codebase. It is uncommon for a feature to be restricted to a single new function (of which you could analyze the use). Adding a feature often means making changes throughout part of the system: changing the semantics of a class or function slightly, adding a parameter here, an extra conditional there, etc. And those changes in turn affect the evolution of the design.

**Comment by queso** [comment]
While I appreciate the debate going on here, I would appreciate keeping the topic around the pull request at hand so we can tweak and agree on a small first step to get things moving forward.

**Comment by queso** [comment]
@martijnwalraven, I tried to address your concerns that this section feels like a promise with these words:

> Including the people above is no guarantee that you will get a response, or ultimately that your pull request will be accepted. This section exists to give some minor guidance on internal Meteor Development Group team structures.

**Comment by martijnwalraven** [comment]
@queso: :+1:

**Comment by queso** [comment]
@martijnwalraven does that mean you are willing to sign off on it then?

**Comment by avital** [comment]
I'm interested in this discussion.

@mitar: When you say, "Many features wanted by community was not accepted into node.js." -- do you mean that those same features were accepted into io.js? Can you point to a few examples?

I am in general agreement with @martijnwalraven. We can definitely not "just merge" things that pass tests. There are so many reasons this is a bad idea including making it harder for people to learn the codebase, making every subsequent change take longer because you have to think about the new case introduced, increasing the likelihood of bugs, etc.

And no, we can't "just remove" a feature once we've shipped it -- that requires a large amount of work, if at all possible. And that also will lead to community backlash, the exact thing we're trying to solve here.

That's the whole point of "designing software". Meteor *should* be designed, not evolved by the sum of all pull requests.

So I generally think that:
1. A bugfix that comes with tests and a convincing explanation for how it doesn't change any other behavior should be merged.
2. New feature suggestions must go through some kind of design review process. And this process may very well be hard. That may be a good thing.

**Comment by avital** [comment]
I'm fine with merging the PR though :)

**Comment by mitar** [comment]
> When you say, "Many features wanted by community was not accepted into node.js." -- do you mean that those same features were accepted into io.js? Can you point to a few examples?

That was the main issue why the fork occurred. Now with merger, those features did get back into the node.js though. You can also notice that they started to increase the version number much quicker now.

io.js FAQ is an useful read: https://iojs.org/en/faq.html

Also: https://github.com/nodejs/node/blob/master/GOVERNANCE.md#readme

Also interesting idea is their approach to long-term support releases: https://github.com/nodejs/LTS/

I think they had very similar issues to what Meteor has: how to combine fast-paced changes and development some want, with stability and production-readiness. (And I know that I am speaking sometimes on both sides of this debate: I would like both long-term stability for some of my projects (like having Blaze around) but also new features in core  to try new things in libraries I do.)

About why fork: http://anandmanisankar.com/posts/nodejs-iojs-why-the-fork/ (Compare decline in [Meteor commits](https://github.com/meteor/meteor/graphs/contributors) with graph of decline in node.js commits - here you see what means "features not getting in". It is even worse at node.js time.)

Features io.js added immediately (and community wanted): new V8 engine, ES6, interface with V8. (I would bring some similarity to how Meteor is still on very old version of underscore. Or node.js engine itself. Or even Mongo - still 2.6?)

See from: https://www.quora.com/Why-is-NodeJS-splitting-into-two-versions-io-js-and-node-js

> IOjs is forking NodeJS because of issues with open governance. The main problem is that NodeJS can't keep up with releases and fixing issues because nobody's really allowing things to get fixed, it isn't as open as things claim. Since V8 is coming along, it would be nice to stay on top of the new changes which is the point of IOjs.

**Comment by mitar** [comment]
> We can definitely not "just merge" things that pass tests.

Do not simplify things I said. This is argumentation ad absurdum. What I said, that we should **faster** merge **feature requests** which (but not merge only those, but for feature requests, we should):
* pass tests
* have documentation
* do not introduce any (visible, clear) backwards incompatibilities (so no intentional breaking of compatibility)
* and some community members say would be useful to them (even if they are not big companies)

**Comment by mitar** [comment]
> And that also will lead to community backlash, the exact thing we're trying to solve here.

Not if you make an open and transparent process, with multiple versions of deprecation. I do not understand how Django can pull this of, but you cannot? :-) It is really simple: you announce publicly that you are considering something to be removed, you make maybe an open vote (like I proposed), you start printing warnings, and then in some later version you remove it.

In general with open governance you will not get a backslash because people will know what is happening. Currently the problem is that a lot of Meteor decisions what gets added and removed is simply done internally. Trello roadmap and votes there does not feel like voting at the moment.

**Comment by mitar** [comment]
> New feature suggestions must go through some kind of design review process. And this process may very well be hard. That may be a good thing.

What about features which are not really features, but like exposing some internals? Or just cleanup of code which makes it easier to interface with. Most of my pull requests are that. :-)

**Comment by RobertLowe** [comment]
@avital

>  ...We can definitely not "just merge" things that pass tests. There are so many reasons this is a bad idea including making it harder for people to learn the codebase

One shouldn't have to learn the entirety of the codebase to make changes efficiently. The area one is working on should be isolated enough to work on. I mean, I shouldn't need to learn `minimongo` to work on `spiderable` for example, the scope is normally smaller then you are purporting.

> ...making every subsequent change take longer because you have to think about the new case introduced, increasing the likelihood of bugs, etc.

This is one of the values provided by testing to prevent regressions and allow for new edge cases to be asserted, as well as help with refactorings. If this is difficult it would because of the current state of testing Meteor itself, which is known to be difficult.

**Comment by martijnwalraven** [comment]
@queso: Yes, I'm ok with this PR!

I also think the discussion that has emerged here is a fruitful one to have (although a PR may not be the best place for it). Improving communication between contributors and the core team is a good first step towards improving the development process. I hope talking about these issues will help us better understanding each others' viewpoints.

For me, an important part of describing a a feature is the motivation: What problem are we trying to solve with this? That not only helps us select which features should be added and how to prioritize them, but it also focuses attention on potential alternatives. Sometimes the same thing can be accomplished in a simpler way, or combined with other features as part of a larger design effort.

> What about features which are not really features, but like exposing some internals? Or just cleanup of code which makes it easier to interface with. Most of my pull requests are that. :-)

There are definitely cases where a small change is all that is needed, and we should be able to merge these sooner, but I believe often larger features are hiding in there. I'm not sure this is a good example, but I'm currently looking at some pull requests against `webapp` @mitar rightfully called us out on because we haven't paid attention to them. One way of looking at them is as small changes that make it easier to interface with. But there is a larger picture, a reason why mitar would like to make these changes. In this case, some of these refer to adding support for 'non-main architectures' or 'custom platforms'. Without context, I can only guess at the larger picture (I'm sorry if I missed a previous discussion). The thing is, I'm definitely interested in uncoupling the current notion of platform from Cordova, but I also think doing so will probably involve more than just these changes. And I think having a discussion around that would be interesting because we can learn from each other and maybe come up with a design that solves more than just the problems mitar is trying to solve.

I definitely don't want to single anyone out or imply this is a one way street. The same goes for features we at MDG are working on. We've too often worked on something internally without paying much attention to users' real needs and without giving them the opportunity to influence the design or even collaborate on the implementation. One change we're trying to make is to share design proposals at a much earlier stage (as is done with the Reactive GraphQL work), so we can have a productive discussion of the issues beforehand and make it more of a collaborative effort.

**Comment by mitar** [comment]
Heh, those pull requests are a bit specific. Especially because I cannot share the rest of the code in my Meteor fork where I have those extra platforms added.

> The thing is, I'm definitely interested in uncoupling the current notion of platform from Cordova, but I also think doing so will probably involve more than just these changes.

They are all that I needed to do for adding extra platforms both on client and server. The patches to add them are then pretty straightforward but there is no API added in my pull request. So those pull requests are just more or less cleanups of the code so that diffs are cleaner.

> And I think having a discussion around that would be interesting because then we can learn from each other and maybe come up with a design that solves more than just the problems mitar is trying to solve.

Agreed. But on the other hand also such dirty approach of adding platforms can be a good insight. Something it is simply good enough. We can alter on port things to a proper API. For now I am mostly just proposing some code cleanups and less special cases so that it is easier to add things to lists of platforms and so on.

**Comment by queso** [comment]
Looks like we are awaiting @yyx990803 and @stubailo as the last two to sign off on this PR.

**Comment by queso** [comment]
Also, @stubailo, I notice a lack of some names on the list. Are there more teams that I've missed in the list perhaps? Maybe most of them are on Galaxy now?

**Comment by yyx990803** [comment]
I think @martijnwalraven and @avital has pretty much summed up any concerns I have, and I think it has been addressed in the added section. So signing off!

**Comment by stubailo** [comment]
@queso looks like we are good to go!

There are some other people as well, let's add them later. Mostly I just want to cover the space of PRs, not make an exhaustive catalog of every person that has or will work on Meteor.

So if there is a _topic_ - a part of the framework - that isn't covered, I can help add it. But I don't want to make this a complete list of people that we have to maintain, and that people will start over-analyzing.

**Comment by stubailo** [comment]
Thanks for the work and discussion everyone! I agree that this PR isn't the best place for an in-depth discussion, for a variety of reasons. Let's find another place.

**Comment by RobertLowe** [comment]
@stubailo we can probably close circle by continuing the conversation on the [forum thread](https://forums.meteor.com/t/how-open-source-meteor-really-is-if-pull-requests-are-not-getting-in) or start a new one. :+1:

**Comment by benstr** [comment]
Holy moly! I'm just coming back to check this issue out. For such a small PR things sure got philosophical! :ear: Not sure what to think it all...

Thanks for hashing it out team and getting this merged!

**Comment by queso** [comment]
@stubailo I am getting a sense that we are missing topics that people have been put in charge of. Maybe you can help outline what we are missing here?

**Comment by queso** [comment]
@RobertLowe I am not sure the forum is the most fruitful place to have this discussion.... I find the threads quickly grow off topic, and this already feels amorphous enough without dragging a bunch of extra topics in.

**Comment by mitar** [comment]
Maybe we need a Meteor retreat where also community members would be invited and flown in. :-)

**Comment by RobertLowe** [comment]
@queso fair enough, perhaps a hackpad document?

**Comment by stubailo** [comment]
Honestly I don't want to have discussions anywhere but the forums and github, that's already enough places for me to keep track of. How can we make one of those places work?

**Comment by queso** [comment]
> Honestly I don't want to have discussions anywhere but the forums and github, that's already enough places for me to keep track of. How can we make one of those places work?

Understandable. I think it should stick to Github then. Perhaps we can make a governance type repo and discuss things there? I know that @nick-coe and @martijnwalraven are working on a similar type project, perhaps we can make it a community discussion without the 'passerby' comments that the forum is prone to. I think the current Guide effort proved that method can work...

**Comment by martijnwalraven** [comment]
Yes, I agree a separate repo is probably the way to go. I think the goal there should be to collaborate on specific documents, such as contributor guidelines or a design review process.

I'm a bit preoccupied with finishing the 1.3 release right now, but I should be able to focus more on this right after!

**Comment by mitar** [comment]
Could we simple use https://github.com/MeteorCommunity/discussions/?

**Comment by mitar** [comment]
Could we simple use https://github.com/MeteorCommunity/discussions/?

**Comment by mitar** [comment]
Could we simple use https://github.com/MeteorCommunity/discussions/?

**Comment by queso** [comment]
@mitar I was hoping to use a Meteor created repo and that we could all participate in the conversation and building of the ideas.

**Comment by benstr** [comment]
+1 :+1:  to keep repo under the Meteor github org.
+1 :+1:  for Meteor retreat to include community members (or official Meteor conference with a break-out session)

**Comment by martijnwalraven** [comment]
Sorry, no retreats for now, but I would also prefer an official repo. What about `meteor/community`?

**Comment by queso** [comment]
> +1 :+1: for Meteor retreat to include community members (or official Meteor conference with a break-out session)

We used to have this, it was called Vanguard :laughing:

**Comment by queso** [comment]
> What about meteor/community?

Perfect! Let's do it.

**Comment by mitar** [comment]
What about moving then that `MeteorCommunity/discussions` into `meteor/community`? I think it would be a great gesture.

**Comment by queso** [comment]
I'd rather start fresh and work it just like the guides did. Get some topic
issues going, ping some people and let the discussion flow as we build
outlines.

**Comment by martijnwalraven** [comment]
I agree with @queso. We could definitely move content from `MeteorCommunity/discussions` over when appropriate, but I think we have to be selective in how we approach this so we can have a focused discussion.

Before discussing specific feature requests, what I hope we can do in the new repo is come up with community guidelines and a process for working together. I don't mean for this to be a purely philosophical endeavor, but I think we can work together more productively if we get the prerequisites right first (things like a shared roadmap, good testing infrastructure, a design review process, etc.).

**Comment by mitar** [comment]
OK.

**Comment by queso** [comment]
> Before discussing specific feature requests, what I hope we can do in the new repo is come up with community guidelines and a process for working together.

This is exactly what I envision by following the same guide building process @martijnwalraven. My hope is to kick the entire thing off with a discussion around building guidelines and the collaboration process.

**Comment by martijnwalraven** [comment]
Yes, I think the guide is a really good example to follow!

**Comment by mitar** [comment]
> Yes, I think the guide is a really good example to follow!

Hm hm. There are also some issues there, but let's not nitpick. ;-)

**Comment by stubailo** [comment]
> Hm hm. There are also some issues there, but let's not nitpick. ;-)

If you're referring to the Blaze thing, we're working on it :] but at least we were transparent about it (or tried to be)

**Comment by mitar** [comment]
:-) Hugs.

**Comment by queso** [comment]
@mitar 

> "let's perfect not be the enemy of the good"

> Hm hm. There are also some issues there, but let's not nitpick. ;-)

Haha.

**Comment by mitar** [comment]
:-)

**Pull request merged**